Port Compression Streams setup to JavaScript

stwrt opened the pull request:
#### 731ed0c7a733198e1dcc79125037e96a777fbb61
<pre>
Port Compression Streams setup to JavaScript
<a href="https://bugs.webkit.org/show_bug.cgi?id=242564">https://bugs.webkit.org/show_bug.cgi?id=242564</a>

Reviewed by Brent Fulgham.

Porting compression streams to JavaScript to handle the error checking,
and the setup of the transform algorithm. The actual encoding will
be handled by the C++ code.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/compression/CompressionStream.cpp: Removed.
* Source/WebCore/Modules/compression/CompressionStream.h: Removed.
* Source/WebCore/Modules/compression/CompressionStream.idl:
* Source/WebCore/Modules/compression/CompressionStream.js: Added.
(initializeCompressionStream):
(getter.readable):
(getter.writable):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp: Added.
(WebCore::CompressionStreamEncoder::encode):
(WebCore::CompressionStreamEncoder::flush):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.h: Added.
(WebCore::CompressionStreamEncoder::create):
(WebCore::CompressionStreamEncoder::CompressionStreamEncoder):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.idl: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventTargetFactory.in:

Canonical link: <a href="https://commits.webkit.org/252383@main">https://commits.webkit.org/252383@main</a>
</pre>

youennf commented on Source/WebCore/Modules/compression/CompressionStream.js:
```diff
@@ -0,0 +1,110 @@
+    if (index === -1)
+        @throwTypeError(errorMessage);
+
+
```
Line not needed

youennf commented on Source/WebCore/Modules/compression/CompressionStream.js:
```diff
@@ -0,0 +1,110 @@
+    const errorMessage = "CompressionStream requires a single argument with the value 'deflate', 'deflate-raw', or 'gzip'.";
+
+    // Error Checking
+    if (arguments.length !== 1 || typeof arguments[0] !== "string")
```
I think length < 1 would be better

stwrt replied:
Thanks for catching this 😀

youennf commented on Source/WebCore/Modules/compression/CompressionStream.js:
```diff
@@ -0,0 +1,110 @@
+        return @Promise.@resolve();
+    };
+    const transformAlgorithm = (chunk) => {
+
```
Line not needed

youennf commented on Source/WebCore/Modules/compression/CompressionStream.js:
```diff
@@ -0,0 +1,110 @@
+        return @Promise.@resolve();
+    };
+    const flushAlgorithm = () => {
+
```
Line not needed

youennf commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (C) 2020 Apple Inc. All rights reserved.
```
2022

youennf commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2020 Apple Inc. All rights reserved.
```
2022

youennf commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -0,0 +1,58 @@
+#include "Formats.h"
+
+#include <JavaScriptCore/Forward.h>
+#include <wtf/Ref.h>
```
Ref.h not need if RefPtr is included I think

youennf commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -0,0 +1,58 @@
+    RefPtr<Uint8Array> flush();
+
+private:
+
```
Line not needed

youennf commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -0,0 +1,58 @@
+
+    const Formats::CompressionFormat m_format;
+
+    CompressionStreamEncoder(unsigned char format) 
```
explicit

youennf commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.idl:
```diff
@@ -0,0 +1,35 @@
+] interface CompressionStreamEncoder {
+    constructor(byte format);
+
+    [PrivateIdentifier] Uint8Array encode(DOMString chunk);
```
Does not seem right that encode would take a String& as input.
Probably https://webidl.spec.whatwg.org/#BufferSource would be more appropriate.

stwrt replied:
Still need to fix this.

stwrt replied:
Fixed in latest patch.

youennf commented on Source/WebCore/Modules/compression/CompressionStream.js:
```diff
@@ -0,0 +1,107 @@
+
+    const algorithms = ['gzip', 'deflate', 'deflate-raw'];
+
+    const findAlgorithm = (element) => element === arguments[0].toLowerCase();
```
Should we call toLowerCase only once?

stwrt replied:
You're right. Have fixed it locally.

We are currently doing 3 to lower cases instead of ideally 1.

youennf commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -0,0 +1,46 @@
+#include "config.h"
+#include "CompressionStreamEncoder.h"
+
+#include "BufferSource.h"
```
Not need since it is included in CompressionStreamEncoder.h

youennf commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -0,0 +1,57 @@
+        return adoptRef(*new CompressionStreamEncoder(format));
+    }
+
+    RefPtr<Uint8Array> encode(const BufferSource&& input);
```
input not needed

youennf commented on Source/WebCore/Modules/compression/DecompressionStream.js:
```diff
@@ -0,0 +1,107 @@
+
+    const algorithms = ['gzip', 'deflate', 'deflate-raw'];
+
+    const findAlgorithm = (element) => element === arguments[0].toLowerCase();
```
Ditto for toLowerCase.

stwrt replied:
Fixed like in compression streams. We only do 1 toLowerCase now

youennf commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -0,0 +1,46 @@
+#include "config.h"
+#include "DecompressionStreamDecoder.h"
+
+#include "BufferSource.h"
```
not needed probably

youennf commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -0,0 +1,57 @@
+#include <JavaScriptCore/Forward.h>
+#include <wtf/RefCounted.h>
+#include <wtf/RefPtr.h>
+#include <wtf/text/WTFString.h>
```
WTFString.h not needed

youennf commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -0,0 +1,57 @@
+#include <JavaScriptCore/Forward.h>
+#include <wtf/RefCounted.h>
+#include <wtf/RefPtr.h>
+#include <wtf/text/WTFString.h>
```
WTFString.h not needed

youennf reviewed (approved):
LGTM as well.
We are missing good C++ transform stream support.
I'll add it in the future, hopefully next month and we will be able to easily migrate these classes, as well as text codec streams.

stwrt committed:
Port Compression Streams setup to JavaScript
https://bugs.webkit.org/show_bug.cgi?id=242564

Reviewed by Brent Fulgham.

Porting compression streams to JavaScript to handle the error checking,
and the setup of the transform algorithm. The actual encoding will
be handled by the C++ code.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/compression/CompressionStream.cpp: Removed.
* Source/WebCore/Modules/compression/CompressionStream.h: Removed.
* Source/WebCore/Modules/compression/CompressionStream.idl:
* Source/WebCore/Modules/compression/CompressionStream.js: Added.
(initializeCompressionStream):
(getter.readable):
(getter.writable):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp: Added.
(WebCore::CompressionStreamEncoder::encode):
(WebCore::CompressionStreamEncoder::flush):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.h: Added.
(WebCore::CompressionStreamEncoder::create):
(WebCore::CompressionStreamEncoder::CompressionStreamEncoder):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.idl: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventTargetFactory.in:

Canonical link: https://commits.webkit.org/252383@main
